packages/web/src/store/application/ui/buy-audio/sagas.ts

```diff
@@ -0,0 +1,460 @@
+  return result
+}
+
+/**
+ * Creates a transaction to transfer all AUDIO from the root associated token account to the user's userbank
+ * @returns the transaction and the amount being transferred
+ */
+function* getTransferTransaction({
+  mintAddress,
+  fromAccount
+}: {
+  mintAddress: string
+  fromAccount: Keypair
+}) {
+  const account = yield* select(getAccountUser)
+  if (!account || !account.userBank) {
+    throw new Error('User does not have a user bank')
+  }
+  const mintPublicKey = new PublicKey(mintAddress)
+  yield* call(waitForLibsInit)
+  const associatedTokenAccount = (yield* call(
+    [
+      window.audiusLibs.solanaWeb3Manager,
+      window.audiusLibs.solanaWeb3Manager.findAssociatedTokenAddress
+    ],
+    fromAccount.publicKey.toString()
+  )) as PublicKey
+  const tokenAccountInfo = (yield* call(
+    [
+      window.audiusLibs.solanaWeb3Manager,
+      window.audiusLibs.solanaWeb3Manager.getAssociatedTokenAccountInfo
+    ],
+    associatedTokenAccount.toString()
+  )) as AccountInfo
+  if (!tokenAccountInfo) {
+    throw new Error('No $AUDIO account in root wallet')
+  }
+  const instruction = Token.createTransferCheckedInstruction(
+    TOKEN_PROGRAM_ID,
+    associatedTokenAccount,
+    mintPublicKey,
+    new PublicKey(account.userBank),
+    fromAccount.publicKey,
+    [],
+    tokenAccountInfo.amount,
+    8
+  )
+  const tx = new Transaction()
+  tx.add(instruction)
+  return { tx, amount: tokenAccountInfo.amount }
+}
+
+/**
+ * Executes a Jupiter Swap given the RouteInfo, account, and the transactionHandler
+ */
+function* doSwap({
+  route,
+  account,
+  transactionHandler
+}: {
+  route: RouteInfo
+  account: Keypair
+  transactionHandler: TransactionHandler
+}) {
+  const jup = yield* call(initJupiterIfNecessary)
+  const {
+    transactions: { setupTransaction, swapTransaction, cleanupTransaction }
+  } = yield* call([jup, jup.exchange], {
+    routeInfo: route,
+    userPublicKey: account.publicKey
+  })
+  if (setupTransaction) {
+    yield* call(
+      sendTransaction,
+      'Setup',
+      setupTransaction,
+      account,
+      transactionHandler
+    )
+  }
+  yield* call(
+    sendTransaction,
+    'Swap',
+    swapTransaction,
+    account,
+    transactionHandler
+  )
+  if (cleanupTransaction) {
+    yield* call(
+      sendTransaction,
+      'Cleanup',
+      cleanupTransaction,
+      account,
+      transactionHandler
+    )
+  }
+}
+
+/**
+ * Swaps the input token (eg. SOL) to the output token (AUDIO) and, if necessary,
+ * transfers resulting AUDIO into the user's userbank from their root Solana account
+ *
+ * Currently only supports SOL => AUDIO
+ *
+ * @returns the amount of AUDIO transferred to the userbank
+ */
+function* doExchange({
+  inputAmount,
+  inputTokenSymbol,
+  outputTokenSymbol,
+  slippage
+}: ReturnType<typeof exchange>['payload']) {
+  yield* call(waitForLibsInit)
+  const connection: Connection = window.audiusLibs.solanaWeb3Manager.connection
```

Comment:
I hate depending on the `window.audiusLibs` here - is there another way? 🥺

Reply:
hmmm can have a think, eventually we will need a way not to for native as well... i guess ideally we have audiusLibs exported from a file?

Reply:
moved most of it to audius-backend